src/slider-new/SliderState.ts

```diff
@@ -0,0 +1,323 @@
+      isDraggingsRef.current,
+      index,
+      dragging,
+    );
+
+    setDraggings(isDraggingsRef.current);
+
+    // Call onChangeEnd if no handles are dragging.
+    if (!wasDragging && isDraggingsRef.current.every(Boolean)) {
+      onChangeStart?.(valuesRef.current);
+    }
+
+    // Call onChangeEnd if no handles are dragging.
+    if (wasDragging && !isDraggingsRef.current.some(Boolean)) {
+      onChangeEnd?.(valuesRef.current);
+    }
+  }
+
+  // Get and Set values based on the index
+  function getThumbValue(index: number) {
+    return values[index];
+  }
+
+  function setThumbValue(index: number, value: number) {
+    if (isDisabled || !isThumbEditable(index)) {
+      return;
+    }
+
+    const thisMin = getThumbMinValue(index);
+    const thisMax = getThumbMaxValue(index);
+
+    // Round value to multiple of step, clamp value between min and max
+    value = clamp(getRoundedValue(value), thisMin, thisMax);
+    valuesRef.current = replaceIndex(valuesRef.current, index, value);
+    setValues(valuesRef.current);
+  }
+
+  function setThumbPercent(index: number, percent: number) {
+    setThumbValue(index, getPercentValue(percent));
+  }
+
+  function getValuePercent(value: number) {
+    return (value - min) / (max - min);
+  }
+
+  function getThumbPercent(index: number) {
+    return getValuePercent(values[index]);
+  }
+
+  function getRoundedValue(value: number) {
+    return Math.round((value - min) / step) * step + min;
+  }
+
+  function getPercentValue(percent: number) {
+    const val = percent * (max - min) + min;
+    return clamp(getRoundedValue(val), min, max);
+  }
+
+  function getThumbMinValue(index: number) {
+    return index === 0 ? min : values[index - 1];
+  }
+
+  function getThumbMaxValue(index: number) {
+    return index === values.length - 1 ? max : values[index + 1];
+  }
+
+  const formatter = useNumberFormatter(formatOptions);
+
+  function getFormattedValue(value: number) {
+    return formatter.format(value);
+  }
+
+  function getThumbValueLabel(index: number) {
+    return getFormattedValue(values[index]);
+  }
+
+  return {
+    values,
+    min,
+    max,
+    step,
+    isDisabled,
+    orientation,
+    trackRef,
+    focusedThumb,
+    setFocusedThumb,
+    getThumbValue,
+    setThumbValue,
+    setThumbPercent,
+    getThumbPercent,
+    getValuePercent,
+    getThumbMinValue,
+    getThumbMaxValue,
+    getPercentValue,
+    getFormattedValue,
+    getThumbValueLabel,
+    isThumbEditable,
+    setThumbEditable,
+    isThumbDragging,
+    setThumbDragging,
+    inputs: inputs.items,
+    registerInput: inputs.registerItem,
+    unregisterInput: inputs.unregisterItem,
+  };
+}
+
+function replaceIndex<T>(array: T[], index: number, value: T) {
+  if (array[index] === value) {
+    return array;
+  }
+
+  return [...array.slice(0, index), value, ...array.slice(index + 1)];
+}
+
+function clampValues(values: number[] | undefined, min: number, max: number) {
```

Comment:
omg, another one. this one is bit different i think we should name it something else, "bulkClamp" or something like that